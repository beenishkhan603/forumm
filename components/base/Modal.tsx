import { useTheme } from '@libs/useTheme'
import { AnimatePresence, motion } from 'framer-motion'
import React, { useEffect, useRef, useState } from 'react'
import ReactDOM from 'react-dom'
import { IoClose } from 'react-icons/io5'
import Box from './Box'
import Text from './Text'

export interface ModalProps {
  /**
   * Children to enclose in the wrapper.
   */
  children: React.ReactNode
  /**
   * Should the header be shown?
   */
  title?: string

  show: boolean

  setShow?: (show: boolean) => void

  closeButton?: boolean

  animating?: (animating: boolean) => void

  version?: boolean

  className?: string

  bgColor?: string

  zIndex?: string
}

const Modal = ({
  children,
  closeButton = true,
  setShow,
  title,
  show,
  animating,
  version = false,
  className = undefined,
  bgColor = undefined,
  zIndex = '99',
}: ModalProps) => {
  const [mounted, setMounted] = useState(false)
  const [maxHeight, setMaxHeight] = useState(window.innerHeight)
  const ref = useRef<Element | null>(null)
  const { theme } = useTheme()
  const isLight = theme.type === 'LIGHT'

  const handleEsc = (event: KeyboardEvent) => {
    if (event.key === 'Escape') {
      if (setShow) setShow(false)
    }
  }

  useEffect(() => {
    ref.current = document.getElementById('modal-root')
    setMounted(true)
    window.addEventListener('keydown', handleEsc)
    return () => {
      window.removeEventListener('keydown', handleEsc)
    }
  }, [])

  useEffect(() => {
    const handleResize = () => {
      setMaxHeight(window.innerHeight)
    }
    window.addEventListener('resize', handleResize)
    return () => {
      window.removeEventListener('resize', handleResize)
    }
  }, [])

  const toggleAnimation = (override: boolean) => {
    if (!animating) return
    animating(override)
  }

  useEffect(() => {
    if (show) {
      document.body.style.overflow = 'hidden'
    } else {
      document.body.style.overflow = ''
    }

    return () => {
      document.body.style.overflow = ''
    }
  }, [show])

  if (!mounted) {
    return null
  }

  return ReactDOM.createPortal(
    // Started causing a build error out of the blue, works fine.
    // @ts-ignore
    <AnimatePresence>
      {show && (
        <motion.div
          /* istanbul ignore next */
          onAnimationComplete={() => toggleAnimation(false)}
          onAnimationStart={() => toggleAnimation(true)}
          initial={{ backgroundColor: '#00000000' }}
          animate={{ backgroundColor: '#00000080' }}
          id="modal"
          transition={{
            duration: 0.3,
          }}
          exit={{ backgroundColor: '#00000000' }}
          className="fixed w-full h-full top-0 left-0 bg-black z-40 flex items-center justify-center"
          style={{ zIndex }}
          onClick={(e) => {
            e.stopPropagation()
            if (setShow) setShow(false)
          }}
        >
          {!version ? (
            <motion.div
              initial={{
                scale: 0,
                opacity: 0,
              }}
              animate={{
                scale: 1,
                opacity: 1,
              }}
              transition={{
                duration: 0.3,
              }}
              exit={{
                scale: 0,
                opacity: 0,
              }}
              className={`relative flex items-center flex-col top-0 w-full mx-2 xs:mt-[155px] md:mt-2 sm:mx-0 !sm:w-5/6 md:min-w-[90%] lg:min-w-[1000px] xl:min-w-[850px] rounded-2xl p-3 !pt-0 sm:p-6 max-w-xl z-50 !max-h-[calc(100vh-155px)] ${
                !isLight
                  ? 'border border-panel-gray'
                  : 'border border-forumm-menu-border'
              } overflow-y-auto ${className ?? ''}`}
              onClick={(e) => e.stopPropagation()}
              style={{
                background: bgColor || theme.backgroundColour,
              }}
            >
              <Box
                className={`sticky top-0 flex w-full items-center py-3 box-border min-h-16`}
                style={{
                  backgroundColor: theme.backgroundColour,
                  zIndex: 1000,
                }}
              >
                <Text
                  className={`text-center text-lg flex-1 font-semibold`}
                  ignoreTheme
                >
                  {title}
                </Text>
                {closeButton && (
                  <button
                    onClick={(e) => {
                      setShow!(false)
                      e.stopPropagation()
                    }}
                    className="absolute right-0"
                  >
                    <IoClose className="w-8 h-8" />
                  </button>
                )}
              </Box>
              <Box className={`w-full`} ignoreTheme>
                {children}
              </Box>
            </motion.div>
          ) : (
            <motion.div
              initial={{
                scale: 0,
                opacity: 0,
              }}
              animate={{
                scale: 1,
                opacity: 1,
              }}
              transition={{
                duration: 0.3,
              }}
              exit={{
                scale: 0,
                opacity: 0,
              }}
              className={`relative flex items-center flex-col top-0 w-full sm:w-3/4 max-w-xl z-50 rounded-xl overflow-y-auto ${
                className ? className : ''
              }`}
              onClick={(e) => e.stopPropagation()}
              style={{
                background: theme.backgroundColour,
                maxHeight: `${maxHeight}px`,
              }}
            >
              <Box className={`w-full`} ignoreTheme>
                {children}
              </Box>
            </motion.div>
          )}
        </motion.div>
      )}
    </AnimatePresence>,
    ref.current!
  )
}

export default Modal
