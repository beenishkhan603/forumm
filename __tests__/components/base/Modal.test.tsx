import Modal from '@components/base/Modal'
import { render } from '@testing-library/react'

it.each([false, true])(
  'renders modal unchanged (without close button)',
  (show) => {
    const { container } = render(
      <>
        <div id="modal-root" />
        <Modal closeButton={false} show={show} setShow={() => {}}>
          Lorem Ipsum...
        </Modal>
      </>
    )
    expect(container).toMatchSnapshot()
  }
)

it.each([false, true])(
  'renders modal unchanged (with close button)',
  (show) => {
    const { container } = render(
      <>
        <div id="modal-root" />
        <Modal closeButton={true} show={show} setShow={() => {}}>
          Lorem Ipsum...
        </Modal>
      </>
    )
    expect(container).toMatchSnapshot()
  }
)

it.each([false, true])('renders modal unchanged (with title)', (show) => {
  const { container } = render(
    <>
      <div id="modal-root" />
      <Modal
        closeButton={true}
        show={show}
        title={'Modal Title'}
        setShow={() => {}}
      >
        Lorem Ipsum...
      </Modal>
    </>
  )
  expect(container).toMatchSnapshot()
})

it('SHOULD not call mocked animate fucntion WHEN animating prop is not passed', () => {
  const mock = jest.fn()
  render(
    <>
      <div id="modal-root" />
      <Modal
        closeButton={true}
        show={true}
        title={'Modal Title'}
        setShow={() => {}}
      >
        Lorem Ipsum...
      </Modal>
    </>
  )
  expect(mock).not.toBeCalled()
})

it('SHOULD call mocked animate fucntion WHEN animating prop is passed', () => {
  const mock = jest.fn()
  render(
    <>
      <div id="modal-root" />
      <Modal
        closeButton={true}
        show={true}
        title={'Modal Title'}
        setShow={() => {}}
        animating={mock}
      >
        Lorem Ipsum...
      </Modal>
    </>
  )
  expect(mock).toBeCalledWith(true)
})
