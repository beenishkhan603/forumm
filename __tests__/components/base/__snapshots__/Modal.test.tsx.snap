// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders modal unchanged (with close button) 1`] = `
<div>
  <div
    id="modal-root"
  />
</div>
`;

exports[`renders modal unchanged (with close button) 2`] = `
<div>
  <div
    id="modal-root"
  >
    <div
      class="fixed w-full h-full top-0 left-0 bg-black z-40 flex items-center justify-center"
      id="modal"
      style="z-index: 99; background-color: rgba(0, 0, 0, 0);"
    >
      <div
        class="relative flex items-center flex-col top-0 w-full mx-2 xs:mt-[155px] md:mt-2 sm:mx-0 !sm:w-5/6 md:min-w-[90%] lg:min-w-[1000px] xl:min-w-[1280px] rounded-2xl p-3 !pt-0 sm:p-6 max-w-xl z-50 !max-h-[calc(100vh-155px)] border border-forumm-menu-border overflow-y-auto "
        style="background: rgb(255, 255, 255); opacity: 0; transform: scale(0) translateZ(0);"
      >
        <div
          class="sticky top-0 flex w-full items-center py-3 box-border min-h-16"
          style="background-color: rgb(255, 255, 255); z-index: 1000;"
        >
          <div
            class="text-center text-lg flex-1 font-semibold"
          />
          <button
            class="absolute right-0"
          >
            <svg
              class="w-8 h-8"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M289.94 256l95-95A24 24 0 00351 127l-95 95-95-95a24 24 0 00-34 34l95 95-95 95a24 24 0 1034 34l95-95 95 95a24 24 0 0034-34z"
              />
            </svg>
          </button>
        </div>
        <div
          class="w-full"
        >
          Lorem Ipsum...
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders modal unchanged (with title) 1`] = `
<div>
  <div
    id="modal-root"
  />
</div>
`;

exports[`renders modal unchanged (with title) 2`] = `
<div>
  <div
    id="modal-root"
  >
    <div
      class="fixed w-full h-full top-0 left-0 bg-black z-40 flex items-center justify-center"
      id="modal"
      style="z-index: 99; background-color: rgba(0, 0, 0, 0);"
    >
      <div
        class="relative flex items-center flex-col top-0 w-full mx-2 xs:mt-[155px] md:mt-2 sm:mx-0 !sm:w-5/6 md:min-w-[90%] lg:min-w-[1000px] xl:min-w-[1280px] rounded-2xl p-3 !pt-0 sm:p-6 max-w-xl z-50 !max-h-[calc(100vh-155px)] border border-forumm-menu-border overflow-y-auto "
        style="background: rgb(255, 255, 255); opacity: 0; transform: scale(0) translateZ(0);"
      >
        <div
          class="sticky top-0 flex w-full items-center py-3 box-border min-h-16"
          style="background-color: rgb(255, 255, 255); z-index: 1000;"
        >
          <div
            class="text-center text-lg flex-1 font-semibold"
          >
            Modal Title
          </div>
          <button
            class="absolute right-0"
          >
            <svg
              class="w-8 h-8"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M289.94 256l95-95A24 24 0 00351 127l-95 95-95-95a24 24 0 00-34 34l95 95-95 95a24 24 0 1034 34l95-95 95 95a24 24 0 0034-34z"
              />
            </svg>
          </button>
        </div>
        <div
          class="w-full"
        >
          Lorem Ipsum...
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders modal unchanged (without close button) 1`] = `
<div>
  <div
    id="modal-root"
  />
</div>
`;

exports[`renders modal unchanged (without close button) 2`] = `
<div>
  <div
    id="modal-root"
  >
    <div
      class="fixed w-full h-full top-0 left-0 bg-black z-40 flex items-center justify-center"
      id="modal"
      style="z-index: 99; background-color: rgba(0, 0, 0, 0);"
    >
      <div
        class="relative flex items-center flex-col top-0 w-full mx-2 xs:mt-[155px] md:mt-2 sm:mx-0 !sm:w-5/6 md:min-w-[90%] lg:min-w-[1000px] xl:min-w-[1280px] rounded-2xl p-3 !pt-0 sm:p-6 max-w-xl z-50 !max-h-[calc(100vh-155px)] border border-forumm-menu-border overflow-y-auto "
        style="background: rgb(255, 255, 255); opacity: 0; transform: scale(0) translateZ(0);"
      >
        <div
          class="sticky top-0 flex w-full items-center py-3 box-border min-h-16"
          style="background-color: rgb(255, 255, 255); z-index: 1000;"
        >
          <div
            class="text-center text-lg flex-1 font-semibold"
          />
        </div>
        <div
          class="w-full"
        >
          Lorem Ipsum...
        </div>
      </div>
    </div>
  </div>
</div>
`;
